File: bindings/lua/causalontology/schema.lua
-- schema.lua - validation against spec/schema/*.schema.json.
--
-- A deliberately small interpreter for exactly the JSON Schema keywords the
-- eight Causalontology schemas use: type, const, enum, pattern, required,
-- properties, additionalProperties, items, minItems, minLength, minimum,
-- maximum, oneOf, and local $ref (#/$defs/...).  "format" is treated as an
-- annotation, as the 2020-12 draft does by default.
--
-- Lua has no regex engine, and needs none here: the schemas use only four
-- anchored pattern families, so pattern_matches() below is a small dedicated
-- validator that recognizes each family from the pattern text itself:
--   ^[0-9a-f]{N}$                    a fixed-length lowercase hex string
--   ^pre:[0-9a-f]{64}$               one prefixed 64-hex identifier
--   ^(pre1|pre2|...):[0-9a-f]{64}$   an alternation of prefixes
--   ^[a-z][a-z0-9_]*$                a lowercase snake_case label (this one
--                                    is a valid Lua pattern verbatim)
-- Any pattern outside these families raises, so a future schema change
-- cannot silently pass unvalidated.

local json = require("causalontology.json")

local schema = {}

local SCHEMA_FILES = {
  cro = "cro.schema.json",
  occurrent = "occurrent.schema.json",
  continuant = "continuant.schema.json",
  realizable = "realizable.schema.json",
  assertion = "assertion.schema.json",
  enrichment = "enrichment.schema.json",
  retraction = "retraction.schema.json",
  succession = "succession.schema.json",
}

local cache = {}

-- The spec/schema directory: CAUSALONTOLOGY_SPEC overrides, else a caller
-- (the conformance runner) sets schema.spec_dir before the first load.
schema.spec_dir = nil

local function schema_dir()
  local env = os.getenv("CAUSALONTOLOGY_SPEC")
  if env then return env .. "/schema" end
  if schema.spec_dir then return schema.spec_dir .. "/schema" end
  error("set CAUSALONTOLOGY_SPEC or schema.spec_dir to the spec/ directory", 0)
end

function schema.load_schema(kind)
  local file = SCHEMA_FILES[kind]
  if not file then error("unknown kind: " .. tostring(kind), 0) end
  if not cache[kind] then
    local path = schema_dir() .. "/" .. file
    local f = assert(io.open(path, "rb"), "cannot open schema " .. path)
    local text = f:read("a")
    f:close()
    cache[kind] = json.decode(text)
  end
  return cache[kind]
end

-- Follow local $ref pointers (#/$defs/...) inside the schema document.
local function resolve(node, root)
  while json.is_object(node) and node["$ref"] ~= nil do
    local ref = node["$ref"]
    if ref:sub(1, 2) ~= "#/" then
      error("only local $ref supported: " .. ref, 0)
    end
    local cursor = root
    for part in ref:sub(3):gmatch("[^/]+") do
      cursor = cursor[part]
    end
    node = cursor
  end
  return node
end

-- The dedicated, regex-free matcher for the schema pattern families.
local function pattern_matches(pattern, value)
  -- family 1: ^[0-9a-f]{N}$ - fixed-length lowercase hex
  local n = pattern:match("^%^%[0%-9a%-f%]{(%d+)}%$$")
  if n then
    return #value == tonumber(n) and value:match("^[0-9a-f]*$") ~= nil
  end
  -- families 2 and 3: ^pre:[0-9a-f]{64}$ / ^(pre1|pre2):[0-9a-f]{64}$
  local prefixes = pattern:match("^%^%(([%w|]+)%):%[0%-9a%-f%]{64}%$$")
                or pattern:match("^%^(%w+):%[0%-9a%-f%]{64}%$$")
  if prefixes then
    local scheme, hex = value:match("^(%w+):([0-9a-f]*)$")
    if not scheme or #hex ~= 64 then return false end
    for candidate in prefixes:gmatch("[^|]+") do
      if candidate == scheme then return true end
    end
    return false
  end
  -- family 4: ^[a-z][a-z0-9_]*$ - a valid Lua pattern verbatim
  if pattern == "^[a-z][a-z0-9_]*$" then
    return value:match("^[a-z][a-z0-9_]*$") ~= nil
  end
  error("unsupported schema pattern: " .. pattern, 0)
end

local function type_ok(value, t)
  if t == "object" then return json.is_object(value) end
  if t == "array" then return json.is_array(value) end
  if t == "string" then return type(value) == "string" end
  if t == "number" then return type(value) == "number" end
  if t == "boolean" then return type(value) == "boolean" end
  error("unsupported schema type: " .. tostring(t), 0)
end

local check  -- forward declaration

check = function(value, node, root, path, errors)
  node = resolve(node, root)

  if node["oneOf"] ~= nil then
    local passing = 0
    for _, sub in ipairs(node["oneOf"]) do
      local suberrs = {}
      check(value, sub, root, path, suberrs)
      if #suberrs == 0 then passing = passing + 1 end
    end
    if passing ~= 1 then
      errors[#errors + 1] = string.format(
        "%s: matches %d of the oneOf branches (need exactly 1)", path, passing)
    end
    return
  end

  local t = node["type"]
  if t ~= nil then
    if not type_ok(value, t) then
      errors[#errors + 1] = string.format("%s: expected %s", path, t)
      return
    end
  end

  if node["const"] ~= nil and not json.deep_equal(value, node["const"]) then
    errors[#errors + 1] = string.format(
      "%s: must equal %s", path, tostring(node["const"]))
  end
  if node["enum"] ~= nil then
    local found = false
    for _, candidate in ipairs(node["enum"]) do
      if value == candidate then found = true break end
    end
    if not found then
      errors[#errors + 1] = string.format(
        "%s: %s not in enumeration", path, tostring(value))
    end
  end
  if node["pattern"] ~= nil and type(value) == "string" then
    if not pattern_matches(node["pattern"], value) then
      errors[#errors + 1] = string.format(
        "%s: %q does not match %s", path, value, node["pattern"])
    end
  end
  if node["minLength"] ~= nil and type(value) == "string" then
    if #value < node["minLength"] then
      errors[#errors + 1] = string.format("%s: shorter than minLength", path)
    end
  end
  if node["minimum"] ~= nil and type(value) == "number" then
    if value < node["minimum"] then
      errors[#errors + 1] = string.format(
        "%s: below minimum %s", path, tostring(node["minimum"]))
    end
  end
  if node["maximum"] ~= nil and type(value) == "number" then
    if value > node["maximum"] then
      errors[#errors + 1] = string.format(
        "%s: above maximum %s", path, tostring(node["maximum"]))
    end
  end

  if json.is_array(value) then
    if node["minItems"] ~= nil and #value < node["minItems"] then
      errors[#errors + 1] = string.format(
        "%s: fewer than %d items", path, node["minItems"])
    end
    if node["items"] ~= nil then
      for i, item in ipairs(value) do
        -- 0-based item paths, matching the Python binding's messages
        check(item, node["items"], root,
              string.format("%s[%d]", path, i - 1), errors)
      end
    end
  end

  if json.is_object(value) then
    local props = node["properties"] or json.new_object()
    if node["required"] ~= nil then
      for _, req in ipairs(node["required"]) do
        if value[req] == nil then
          errors[#errors + 1] = string.format(
            "%s: required property '%s' missing", path, req)
        end
      end
    end
    if node["additionalProperties"] == false then
      for _, key in ipairs(json.keys(value)) do
        if props[key] == nil then
          errors[#errors + 1] = string.format(
            "%s: additional property '%s'", path, key)
        end
      end
    end
    for _, key in ipairs(json.keys(props)) do
      if value[key] ~= nil then
        check(value[key], props[key], root, path .. "." .. key, errors)
      end
    end
  end
end

-- (ok, reasons) - structural validity against the kind's JSON Schema.
function schema.validate_schema(obj, kind)
  local canonical = require("causalontology.canonical")
  kind = kind or canonical.infer_kind(obj)
  local root = schema.load_schema(kind)
  local errors = {}
  check(obj, root, root, "$", errors)
  return #errors == 0, errors
end

return schema

File: bindings/haskell/src/Causalontology/Schema.hs
-- | Schema validation against spec\/schema\/*.schema.json.
--
-- A deliberately small interpreter for exactly the JSON Schema keywords
-- the eight Causalontology schemas use: type, const, enum, pattern,
-- required, properties, additionalProperties, items, minItems, minLength,
-- minimum, maximum, oneOf, and local $ref (#\/$defs\/...). @format@ is
-- treated as an annotation, as the 2020-12 draft does by default.
--
-- The @pattern@ keyword is served by a tiny backtracking matcher
-- ('patternMatch') covering exactly the regular-expression subset those
-- schemas use: literals, @^@ and @$@ anchors, character classes with
-- ranges, alternation groups, @*@, and @{n}@ repetition.
module Causalontology.Schema
  ( Schemas
  , schemaFileTable
  , loadSchemas
  , validateSchema
  , patternMatch
  ) where

import Causalontology.Canonical (inferKind)
import Causalontology.Json
import qualified Data.ByteString as B
import Data.Char (isDigit)
import Data.List (isPrefixOf, tails)
import System.FilePath ((</>))

-- | The eight loaded schemas, keyed by kind.
type Schemas = [(String, JValue)]

-- | Kind to schema file name.
schemaFileTable :: [(String, String)]
schemaFileTable =
  [ ("causal_relation_object", "cro.schema.json")
  , ("occurrent", "occurrent.schema.json")
  , ("continuant", "continuant.schema.json")
  , ("realizable", "realizable.schema.json")
  , ("assertion", "assertion.schema.json")
  , ("enrichment", "enrichment.schema.json")
  , ("retraction", "retraction.schema.json")
  , ("succession", "succession.schema.json")
  ]

-- | Load all eight schemas from a schema directory (spec\/schema).
loadSchemas :: FilePath -> IO Schemas
loadSchemas schemaDir = mapM loadOne schemaFileTable
  where
    loadOne (kind, file) = do
      bytes <- B.readFile (schemaDir </> file)
      case parseJson (utf8Decode (B.unpack bytes)) of
        Right v -> return (kind, v)
        Left err -> ioError (userError ("cannot parse " ++ file ++ ": " ++ err))

-- | @(ok, reasons)@ - structural validity against the kind's JSON Schema.
-- The outer 'Left' fires only for an unknown or uninferable kind.
validateSchema :: Schemas -> JValue -> Maybe String -> Either String (Bool, [String])
validateSchema schemas obj mkind = do
  kind <- case mkind of
    Just k -> Right k
    Nothing -> inferKind obj
  root <- case lookup kind schemas of
    Just r -> Right r
    Nothing -> Left ("unknown kind: " ++ kind)
  let errs = check obj root root "$"
  Right (null errs, errs)

-- | Follow local @$ref@ chains to the referenced schema node.
resolveRef :: JValue -> JValue -> JValue
resolveRef schema root = case objGet "$ref" schema of
  Just (JStr ref)
    | "#/" `isPrefixOf` ref ->
        let parts = splitOn '/' (drop 2 ref)
            node = foldl (\n part -> maybe JNull id (objGet part n)) root parts
        in resolveRef node root
  Just _ -> error "only local $ref supported"
  Nothing -> schema

-- | Split a string on a separator character.
splitOn :: Char -> String -> [String]
splitOn sep s = case break (== sep) s of
  (piece, []) -> [piece]
  (piece, _ : rest) -> piece : splitOn sep rest

-- | A Python-repr-flavoured rendering for error messages.
jRepr :: JValue -> String
jRepr (JStr s) = "'" ++ s ++ "'"
jRepr (JInt n) = show n
jRepr (JFloat f) = show f
jRepr (JBool b) = if b then "True" else "False"
jRepr JNull = "None"
jRepr (JArr xs) = "[" ++ concatMap (\x -> jRepr x ++ ", ") xs ++ "]"
jRepr (JObj kvs) = "{" ++ concatMap (\(k, v) -> "'" ++ k ++ "': " ++ jRepr v ++ ", ") kvs ++ "}"

-- | The recursive keyword interpreter; returns the error list.
check :: JValue -> JValue -> JValue -> String -> [String]
check value schema0 root path =
  case objGet "oneOf" schema of
    Just (JArr subs) ->
      let passing = length [ () | sub <- subs, null (check value sub root path) ]
      in [ path ++ ": matches " ++ show passing ++ " of the oneOf branches (need exactly 1)"
         | passing /= 1
         ]
    _ -> case typeError of
      Just err -> [err]
      Nothing ->
        concat
          [ constErrors
          , enumErrors
          , patternErrors
          , minLengthErrors
          , minimumErrors
          , maximumErrors
          , arrayErrors
          , objectErrors
          ]
  where
    schema = resolveRef schema0 root

    typeError = case objGet "type" schema of
      Just (JStr t)
        | not (typeMatches t) -> Just (path ++ ": expected " ++ t)
      _ -> Nothing

    typeMatches t = case (t, value) of
      ("object", JObj _) -> True
      ("array", JArr _) -> True
      ("string", JStr _) -> True
      ("boolean", JBool _) -> True
      ("number", JInt _) -> True
      ("number", JFloat _) -> True
      _ -> False

    constErrors = case objGet "const" schema of
      Just c | not (jEqual value c) -> [path ++ ": must equal " ++ jRepr c]
      _ -> []

    enumErrors = case objGet "enum" schema of
      Just (JArr items)
        | not (any (jEqual value) items) ->
            [path ++ ": " ++ jRepr value ++ " not in enumeration"]
      _ -> []

    patternErrors = case (objGet "pattern" schema, value) of
      (Just (JStr pat), JStr s)
        | not (patternMatch pat s) ->
            [path ++ ": " ++ jRepr value ++ " does not match " ++ pat]
      _ -> []

    minLengthErrors = case (objGet "minLength" schema >>= jNumber, value) of
      (Just n, JStr s)
        | fromIntegral (length s) < n -> [path ++ ": shorter than minLength"]
      _ -> []

    minimumErrors = case (objGet "minimum" schema, jNumber value) of
      (Just bound, Just v)
        | Just b <- jNumber bound
        , v < b ->
            [path ++ ": below minimum " ++ jRepr bound]
      _ -> []

    maximumErrors = case (objGet "maximum" schema, jNumber value) of
      (Just bound, Just v)
        | Just b <- jNumber bound
        , v > b ->
            [path ++ ": above maximum " ++ jRepr bound]
      _ -> []

    arrayErrors = case value of
      JArr xs ->
        let minItemsErrs = case objGet "minItems" schema >>= jNumber of
              Just n
                | fromIntegral (length xs) < n ->
                    [path ++ ": fewer than " ++ show (truncate n :: Integer) ++ " items"]
              _ -> []
            itemErrs = case objGet "items" schema of
              Just sub ->
                concat
                  [ check item sub root (path ++ "[" ++ show i ++ "]")
                  | (i, item) <- zip [0 :: Int ..] xs
                  ]
              Nothing -> []
        in minItemsErrs ++ itemErrs
      _ -> []

    objectErrors = case value of
      JObj kvs ->
        let props = maybe [] objEntries (objGet "properties" schema)
            requiredErrs =
              [ path ++ ": required property '" ++ r ++ "' missing"
              | r <- maybe [] strList (objGet "required" schema)
              , not (any ((== r) . fst) kvs)
              ]
            additionalErrs = case objGet "additionalProperties" schema of
              Just (JBool False) ->
                [ path ++ ": additional property '" ++ k ++ "'"
                | (k, _) <- kvs
                , not (any ((== k) . fst) props)
                ]
              _ -> []
            propErrs =
              concat
                [ check v sub root (path ++ "." ++ k)
                | (k, sub) <- props
                , Just v <- [lookup k kvs]
                ]
        in requiredErrs ++ additionalErrs ++ propErrs
      _ -> []

-- ---------------------------------------------------------------------------
-- the pattern subset
-- ---------------------------------------------------------------------------

-- | One regex atom.
data RxAtom
  = RxLit Char
  | RxClass [(Char, Char)]
  | RxGroup [[RxItem]]

-- | An atom with its quantifier.
data RxItem = RxItem RxAtom RxQuant

-- | The supported quantifiers.
data RxQuant = QOne | QStar | QRep Int

-- | Python @re.search@ semantics over the schema pattern subset: does
-- the pattern match anywhere in the string (honouring @^@ and @$@)?
patternMatch :: String -> String -> Bool
patternMatch pat s =
  case parseAlts body of
    Nothing -> False
    Just alts ->
      let candidates = if anchoredStart then [s] else tails s
          accepts rest = not anchoredEnd || null rest
      in any
           (\start -> any accepts (concatMap (\alt -> matchItems alt start) alts))
           candidates
  where
    (anchoredStart, p1) = case pat of
      '^' : r -> (True, r)
      _ -> (False, pat)
    (anchoredEnd, body) =
      if not (null p1) && last p1 == '$'
        then (True, init p1)
        else (False, p1)

-- | Parse a top-level alternation (the whole remaining pattern).
parseAlts :: String -> Maybe [[RxItem]]
parseAlts str = do
  (alt1, rest) <- parseSeq str
  case rest of
    '|' : more -> do
      alts <- parseAlts more
      Just (alt1 : alts)
    "" -> Just [alt1]
    _ -> Nothing

-- | Parse a sequence of quantified atoms, stopping at @|@, @)@, or the end.
parseSeq :: String -> Maybe ([RxItem], String)
parseSeq str = case str of
  "" -> Just ([], "")
  '|' : _ -> Just ([], str)
  ')' : _ -> Just ([], str)
  _ -> do
    (atom, r1) <- parseAtom str
    (quant, r2) <- parseQuant r1
    (restItems, r3) <- parseSeq r2
    Just (RxItem atom quant : restItems, r3)

-- | Parse one atom.
parseAtom :: String -> Maybe (RxAtom, String)
parseAtom ('(' : r) = do
  (alts, r') <- parseGroupAlts r
  Just (RxGroup alts, r')
parseAtom ('[' : r) = do
  (ranges, r') <- parseClass r []
  Just (RxClass ranges, r')
parseAtom ('\\' : c : r) = Just (RxLit c, r)
parseAtom (c : r)
  | c `notElem` "|)*{" = Just (RxLit c, r)
parseAtom _ = Nothing

-- | Parse the alternatives of a group up to the closing parenthesis.
parseGroupAlts :: String -> Maybe ([[RxItem]], String)
parseGroupAlts str = do
  (alt1, rest) <- parseSeq str
  case rest of
    '|' : more -> do
      (alts, r') <- parseGroupAlts more
      Just (alt1 : alts, r')
    ')' : more -> Just ([alt1], more)
    _ -> Nothing

-- | Parse a character class body up to the closing bracket.
parseClass :: String -> [(Char, Char)] -> Maybe ([(Char, Char)], String)
parseClass (']' : r) acc = Just (reverse acc, r)
parseClass (a : '-' : b : r) acc
  | b /= ']' = parseClass r ((a, b) : acc)
parseClass (a : r) acc = parseClass r ((a, a) : acc)
parseClass [] _ = Nothing

-- | Parse an optional quantifier.
parseQuant :: String -> Maybe (RxQuant, String)
parseQuant ('*' : r) = Just (QStar, r)
parseQuant ('{' : r) =
  let (ds, r1) = span isDigit r
  in case r1 of
       '}' : r2 | not (null ds) -> Just (QRep (read ds), r2)
       _ -> Nothing
parseQuant r = Just (QOne, r)

-- | All ways a quantified-atom sequence can consume a prefix of the
-- string; each result is the remaining suffix.
matchItems :: [RxItem] -> String -> [String]
matchItems [] str = [str]
matchItems (RxItem atom quant : rest) str = case quant of
  QOne -> continue (matchAtom atom str)
  QRep n -> continue (matchTimes n str)
  QStar -> continue (matchStar str)
  where
    continue nexts = concatMap (matchItems rest) nexts
    matchTimes :: Int -> String -> [String]
    matchTimes 0 s' = [s']
    matchTimes k s' = concatMap (matchTimes (k - 1)) (matchAtom atom s')
    -- the progress guard keeps a hypothetical empty-matching group from
    -- looping; every atom in the schema patterns consumes input
    matchStar s' =
      s' : concatMap matchStar (filter (\x -> length x < length s') (matchAtom atom s'))

-- | All ways one atom can consume a prefix of the string.
matchAtom :: RxAtom -> String -> [String]
matchAtom (RxLit c) (x : xs) | x == c = [xs]
matchAtom (RxClass ranges) (x : xs)
  | any (\(lo, hi) -> lo <= x && x <= hi) ranges = [xs]
matchAtom (RxGroup alts) str = concatMap (\alt -> matchItems alt str) alts
matchAtom _ _ = []
